# Lab dualidad y holgura complementaria

## Sebastian Guayacan Mesa
## Andrés Felipe Cordero Martinez

# Punto 1

In [2]:
from matplotlib import pyplot, pylab, mlab
from pylab import *
from scipy.optimize import linprog
import numpy as np
import math 
from math import factorial
from scipy.optimize import minimize
import sympy as sym
from itertools import combinations
import pandas as pd
import time
from IPython.display import display

In [4]:
def PL (A,b,fun):
    dir4 = soluciones_basicas(A,b)
    # Se inicializa el vector que guardara el tipo de solución que se presenta
    solucion = []
    tip_sol = []
    base = []
    # Se inicializa el vector que guardara el Valor dado por la solución presentada
    val = []
    ns = []
    # Se crea el diccionario que se utilizara para realizar la impresión de la tabla
    rta = {"Solucion": solucion,
            "Base": base,
            "Tipo de solucion": tip_sol,
            "Valor": val
            }
    # Para encontrar el tipo de solución se va a hacer un recorrido por las soluciones basicas encontradas en el anterior literal
    n=0
    for i in dir4["Solucion"]:
        
        agregar = "Solucion basica factible" # El estado por defecto va a ser una solución basica factible
        for j in i: # Se recorre cada valor dado en la solucion para encontrar que cumplan con la condicion de factibilidad, todos deben ser mayor o igual a cero
            if j <= -0.0: # Si se encuentra un valor negativo entrar a cambiar el valor default de solución basica factible por solucion basica no factible
                agregar = "Solucion basica no factible" 
        if agregar == "Solucion basica factible":
            solucion.append(i)
            base.append(dir4["Base"][n])
            tip_sol.append(agregar)
            ns.append(n)
            
        n = n +1

    # Se inicializa la variable que va a recorrer las soluciones basicas encontradas
    d = 0
    # Se crea el vector que va a contener la funcion de costos a evaluar
     
    # Se inicia el recorrido por la bases de las soluciones basicas encontradas
    for i in rta["Base"]:
            anadir = 0 # Se inicializa el valor evaluado que se agregara a la lista de valor obtenida al evaluar la solucion en la función de costos
            f = 0 # Se inicializa la variable que va a recorrer cada valor de la solucion actual
            c = 0 # Se inicializa la variable que va a recorrer la posicion actual de la funcion de costos
            for j in i[0]: # Se recorre el primer vector entregado por la base
                    if j > 0.0:
                            anadir += dir4["Solucion"][(ns[d])][f]*fun[c] # Se agrega valor a la variable que tendra la evaluacion de la función de costos
                            f += 1 # Se avanza una posición en el vector solución
                    c += 1 # Se avanza una posición en la función de costos
                    if c == 3: # Una vez avance por todas las posiciones de la funcion de costos pasar a la siguiente solucion basica
                            break
            val.append(-anadir) # Se agrega la evaluacion de la solucion actual en la funcion de costos a la lista que tiene el valor de la solucion actual en la funcion de costos
             # Este valor se multiplica por - para tener de esta forma un resultado del problema maximizado y no minimizado
            d += 1 # Se avanza a la siguiente solucion basica
        # Se implementa el codigo necesario para usar la función display y mostrar los resultados en la tabla esperada
    

     
    pd.set_option("max_colwidth",100)
    df_dir6 = pd.DataFrame(rta)
    display (df_dir6)
#Esta función fue elaborada para resolver el laboratorio 4 de la clase.
def soluciones_basicas (A,b):
    sol_bas = []
    bases = []
    columnas = np.size(A[0])
    filas = np.size(b[:,0])
    ans = {"Solucion": sol_bas,
            "Base":bases  }
    matrices = list (combinations(np.arange(0, columnas, 1),filas))
    
    for i in range(len(matrices)):
        sub_A = []
        base =np.zeros((filas,columnas))
        for j in ((matrices[i])):
                sub_A.append(A[:,j])
        sub_A =  (np.array(sub_A)).transpose()
        if np.linalg.det(sub_A) !=0:
          aux  = np.linalg.solve(sub_A,b)
          aux = np.round(aux,5)
          base[:,list(matrices[i])]=sub_A 
          bases.append(base)
          sol_bas.append(aux)
          
    return  ans

## Normal

In [25]:
A = np.array([[7/10 ,1,1,0,0,0],[1/2,5/6,0,1,0,0],[1, 2/3, 0,0,1,0], [1/10,1/4,0,0,0,1]])#matriz con los coeficientes de restricciones
b=np.array([[630],[600], [708],[135]]) # vector de restricciones
s = PL(A,b,[-7,-9,0,0,0,0]) #vector con los coeficientes de la funcion a minimizar    

Base  \
0  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
1  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
2  [[0.7, 0.0, 1.0, 0.0, 0.0, 0.0], [0.5, 0.0, 0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0, 0.0, 0.0],...   
3  [[0.0, 1.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [0.0, 0.66666666...   
4  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0],...   

                               Solucion          Tipo de solucion     Valor  
0  [[300.0], [420.0], [100.0], [128.0]]  Solucion basica factible  [5880.0]  
1   [[540.0], [252.0], [120.0], [18.0]]  Solucion basica factible  [6048.0]  
2   [[708.0], [134.4], [246.0], [64.2]]  Solucion basica factible  [4956.0]  
3   [[540.0], [90.0], [150.0], [348.0]]  Solucion basica factible  [4860.0]  
4  [[630.0], [600.0], [708.0], [135.0]]  Solucion basica factible    [-0.0]

La utilidad obtenida al comercializar productos de ambas líneas es de 6048 dólares y se obtiene con 540 productos de la línea estándar y 252 de la línea deluxe.


## Solo vendiendo deluxe:

Se contempla el escenario en el que solo se comercializan unidades de la línea Deluxe.

In [26]:
A = np.array([[0 ,1,1,0,0,0],[0,5/6,0,1,0,0],[0, 2/3, 0,0,1,0], [0,1/4,0,0,0,1]])
#matriz con los coeficientes de restricciones
b=np.array([[630],[600], [708],[135]]) 
 # vector de restricciones
s = PL(A,b,[0,-9,0,0,0,0]) 
#vector con los coeficientes de la funcion a minimizar

Base  \
0  [[0.0, 1.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [0.0, 0.66666666...   
1  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0],...   

                               Solucion          Tipo de solucion     Valor  
0   [[540.0], [90.0], [150.0], [348.0]]  Solucion basica factible  [4860.0]  
1  [[630.0], [600.0], [708.0], [135.0]]  Solucion basica factible    [-0.0]

La utilidad obtenida al comercializar productos de solo la  línea deluxe es de 4860 dólares y se obtiene con 540 productos de la línea deluxe y ninguno de la línea estándar.

## Solo vendiendo estándar:

Así mismo, se contempla la posibilidad de solo vender unidades de la línea estándar


In [27]:
A = np.array([[7/10 ,0,1,0,0,0],[1/2,0,0,1,0,0],[1, 0, 0,0,1,0], [1/10,0,0,0,0,1]])
#matriz con los coeficientes de restricciones
b=np.array([[630],[600], [708],[135]])
 # vector de restricciones
s = PL(A,b,[-7,0,0,0,0,0])
#vector con los coeficientes de la funcion a minimizar

Base  \
0  [[0.7, 0.0, 1.0, 0.0, 0.0, 0.0], [0.5, 0.0, 0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0, 0.0, 0.0],...   
1  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0],...   

                               Solucion          Tipo de solucion     Valor  
0   [[708.0], [134.4], [246.0], [64.2]]  Solucion basica factible  [4956.0]  
1  [[630.0], [600.0], [708.0], [135.0]]  Solucion basica factible    [-0.0]

La utilidad obtenida al comercializar productos de solo la  línea estándar es de 4956 dólares y se obtiene con 708 productos de la línea estándar y ninguno de la línea deluxe.

Al comparar los tres escenarios mencionados se obtiene la siguiente tabla
|Tipo|Valor|Solución|
|:----|:----|:----|
|normal|6048|[540,252]|
|B. estándar|4956|[540,252]|
|B. deluxe|4860|[540,252]|

Lo cual indica que se generan mayores alternativas cuando se producen productos de la línea estándar y deluxe.


## Segundo punto 

El departamento de contabilidad de su empresa le presenta tres estrategias para aumentar la contribución
total a la utilidad por parte de las bolsas de golf.


a) 

Se contempla en escenario de incrementar la utilidad por bolsa para el modelo deluxe a 18 dólares. La utilidad por bolsa del modelo estándar permanece en 7 dólares


In [28]:
A = np.array([[7/10 ,1,1,0,0,0],[1/2,5/6,0,1,0,0],[1, 2/3, 0,0,1,0], [1/10,1/4,0,0,0,1]])
#matriz con los coeficientes de restricciones
b=np.array([[630],[600], [708],[135]])
 # vector de restricciones
#vector con los coeficientes de la funcion a minimizar
s = PL(A,b,[-7,-18,0,0,0,0])

Base  \
0  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
1  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
2  [[0.7, 0.0, 1.0, 0.0, 0.0, 0.0], [0.5, 0.0, 0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0, 0.0, 0.0],...   
3  [[0.0, 1.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [0.0, 0.66666666...   
4  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0],...   

                               Solucion          Tipo de solucion     Valor  
0  [[300.0], [420.0], [100.0], [128.0]]  Solucion basica factible  [9660.0]  
1   [[540.0], [252.0], [120.0], [18.0]]  Solucion basica factible  [8316.0]  
2   [[708.0], [134.4], [246.0], [64.2]]  Solucion basica factible  [4956.0]  
3   [[540.0], [90.0], [150.0], [348.0]]  Solucion basica factible  [9720.0]  
4  [[630.0], [600.0], [708.0], [135.0]]  Solucion basica factible    [-0.0]

La utilidad obtenida al implementar la alternativa a es de 9720 dólares y se obtiene 540 productos de la línea deluxe y ninguno de la línea estándar.

b)  Tamnbién se contempla la posibilidad de incorporar al proceso de producción una materia prima de bajo costo para la fabricación del modelo estándar para que se pueda incrementar su utilidad por bolsa a 14 dólares. La utilidad por bolsa del modelo deluxe permanece en 9 dólares.

In [29]:
A = np.array([[7/10 ,1,1,0,0,0],[1/2,5/6,0,1,0,0],[1, 2/3, 0,0,1,0], [1/10,1/4,0,0,0,1]])
#matriz con los coeficientes de restricciones
b=np.array([[630],[600], [708],[135]])
# vector de restricciones
#vector con los coeficientes de la funcion a minimizar
s = PL(A,b,[-14,-9,0,0,0,0])


Base  \
0  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
1  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
2  [[0.7, 0.0, 1.0, 0.0, 0.0, 0.0], [0.5, 0.0, 0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0, 0.0, 0.0],...   
3  [[0.0, 1.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [0.0, 0.66666666...   
4  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0],...   

                               Solucion          Tipo de solucion     Valor  
0  [[300.0], [420.0], [100.0], [128.0]]  Solucion basica factible  [7980.0]  
1   [[540.0], [252.0], [120.0], [18.0]]  Solucion basica factible  [9828.0]  
2   [[708.0], [134.4], [246.0], [64.2]]  Solucion basica factible  [9912.0]  
3   [[540.0], [90.0], [150.0], [348.0]]  Solucion basica factible  [4860.0]  
4  [[630.0], [600.0], [708.0], [135.0]]  Solucion basica factible    [-0.0]

La utilidad obtenida al implementar la alternativa b es de 9912 dólares y se obtiene 708 productos de la línea estándar y ninguno de la línea deluxe.

c) Finalmente, se planea obtener un equipo de trabajo adicional que aumente la cantidad de horas de trabajo disponible en un 70 % para cada proceso. La utilidad por bolsa de los modelos no cambia, 7 dólares para estándar y 9 dólares para deluxe.


In [30]:
A = np.array([[7/10 ,1,1,0,0,0],[1/2,5/6,0,1,0,0],[1, 2/3, 0,0,1,0], [1/10,1/4,0,0,0,1]])
#matriz con los coeficientes de restricciones
b=np.array([[1071],[1020], [1203.6],[229.5]])
# vector de restricciones
#vector con los coeficientes de la funcion a minimizar
s = PL(A,b,[-7,-9,0,0,0,0])


Base  \
0  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
1  [[0.7, 1.0, 0.0, 0.0, 0.0, 0.0], [0.5, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [1.0, 0.66666666...   
2  [[0.7, 0.0, 1.0, 0.0, 0.0, 0.0], [0.5, 0.0, 0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0, 0.0, 0.0],...   
3  [[0.0, 1.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.8333333333333334, 0.0, 1.0, 0.0, 0.0], [0.0, 0.66666666...   
4  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 0.0],...   

                                  Solucion          Tipo de solucion  \
0     [[510.0], [714.0], [170.0], [217.6]]  Solucion basica factible   
1      [[918.0], [428.4], [204.0], [30.6]]  Solucion basica factible   
2  [[1203.6], [228.48], [418.2], [109.14]]  Solucion basica factible   
3     [[918.0], [153.0], [255.0], [591.6]]  Solucion basica factible   
4  [[1071.0], [1020.0], [1203.6], [229.5]]  Solucion basica factible   

                 Valor  
0             [9996.0]  
1            [10281.6]  
2  [8425.199999999999]  
3             [8262.0]  
4               [-0.0]

Finalmente, la utilidad obtenida al implementar la alternativa c es de 10281 dólares y se obtiene 428 productos de la línea deluxe y 918 de la línea estándar.

## De las tres alternativas planteadas se obtiene que:




|Punto	|Valor|	S_primal|
|:----|:----|:----|
|a|9.720|0,540|
|b |9.912|708,0|
|c | 10.281|918,428|

